### PR TITLE
[mlir] Add missing fields in DICompositeTypeAttr.

### PR DESCRIPTION
**Pull request by abidh**
The fortran arrays use 'dataLocation', 'rank', 'allocated' and 'associated' fields of the DICompositeType. These were not available in 'DICompositeTypeAttr'. This PR adds the missing fields.

### CODE REVIEW

File: mlir/include/mlir-c/Dialect/LLVM.h
Comment (gysit) [inline mlir/include/mlir-c/Dialect/LLVM.h:243]:
```suggestion
    MlirAttribute dataLocation, MlirAttribute rank, MlirAttribute allocated,
```
nit: should it be dataLocation as well for consistency?

File: mlir/lib/CAPI/Dialect/LLVM.cpp
Comment (gysit) [inline mlir/lib/CAPI/Dialect/LLVM.cpp:167]:
```suggestion
    MlirAttribute dataLocation, MlirAttribute rank, MlirAttribute allocated,
```

File: mlir/test/Target/LLVMIR/llvmir-debug.mlir
Comment (gysit) [inline mlir/test/Target/LLVMIR/llvmir-debug.mlir:488]:
```suggestion
// -----

```
nit: missing newline

File: mlir/test/Target/LLVMIR/Import/composite-type-fortran.ll
Comment (gysit) [inline mlir/test/Target/LLVMIR/Import/composite-type-fortran.ll:27]:
```suggestion
```
nit: drop trailing newlines.

File: mlir/test/Target/LLVMIR/llvmir-debug.mlir
Comment (gysit) [inline mlir/test/Target/LLVMIR/llvmir-debug.mlir:522]:
```suggestion

llvm.func @fn_with_composite() {
  llvm.return
} loc(fused<#sp0>["foo.mlir":1:1])

// CHECK: !DICompositeType(
// CHECK-SAME: dataLocation: !DIExpression(DW_OP_push_object_address, DW_OP_deref)
// CHECK-SAME: associated: !DIExpression(DW_OP_lit0, DW_OP_eq)
// CHECK-SAME: allocated: !DIExpression(DW_OP_lit0, DW_OP_ne)
// CHECK-SAME: rank: !DIExpression(DW_OP_push_object_address, DW_OP_plus_uconst, 16, DW_OP_deref)
```

Comment (gysit) [reply mlir/test/Target/LLVMIR/llvmir-debug.mlir:522]: nit: I would check for the composite type itself here but not necessarily for the enclosing subroutine type. Also added some whitespaces to format the function.

**Review by gysit — COMMENTED** [review]
Thanks!

Looks good. I added some minor formatting / naming comments.

File: mlir/test/Target/LLVMIR/llvmir-debug.mlir
Comment (abidh) [reply mlir/test/Target/LLVMIR/llvmir-debug.mlir]: I used this approach initially but observed that test will fail because `FileCheck` will match an earlier `!DICompositeType` in the file. In the updated patch, I used your suggestion but added a `CHECK-LABEL` to avoid the wrong match.

File: mlir/test/Target/LLVMIR/llvmir-debug.mlir
Comment (gysit) [inline mlir/test/Target/LLVMIR/llvmir-debug.mlir:516]:
```suggestion
>

llvm.func @fn_with_composite() {
  llvm.return
} loc(fused<#sp0>["foo.mlir":1:1])

// CHECK-LABEL: define void @fn_with_composite()
```
ultra nit: Added some more spaces here. The CHECK-LABEL makes a lot of sense!

**Review by gysit — APPROVED** [review]
Thanks LGTM modulo last nit!

**Comment by abidh** [comment]
The #92568 introduced call of [DICompositeTypeAttr::get](https://github.com/llvm/llvm-project/blob/1383cb6e648ba9f4516fa2317f948116c4f35c10/flang/lib/Optimizer/Transforms/DebugTypeGenerator.cpp#L72) in flang. That will need to be updated once this PR is merged. I have opened #93516 for that.

**Comment by gysit** [comment]
There seem to be issues with the build bots still. Once they are fixed this commit can be landed!

**Comment by abidh** [comment]
> The #92568 introduced call of [DICompositeTypeAttr::get](https://github.com/llvm/llvm-project/blob/1383cb6e648ba9f4516fa2317f948116c4f35c10/flang/lib/Optimizer/Transforms/DebugTypeGenerator.cpp#L72) in flang

After opening of this PR, #92568 went in which added a call to DICompositeTypeAttr::get in flang that was causing the build failure. I have pushed a commit to adjust the call. I will wait for bots to turn green before landing.

**Pull request merged**
